src/ocispec/ocispec.cpp

```diff
@@ -0,0 +1,184 @@
+    }
+
+    for (size_t i = 0; i < imageSpec.mConfig.mEntryPoint.Size(); i++) {
+        mImageSpec.config.entrypoint[i] = const_cast<char*>(imageSpec.mConfig.mEntryPoint[i].CStr());
+    }
+
+    json_obj_encode_buf(ImageSpecJsonDescr, ARRAY_SIZE(ImageSpecJsonDescr), &mImageSpec, (char*)mJsonFileBuffer.Get(),
+        cJsonMaxContentSize);
+
+    auto err = WriteEncodedJsonBufferToFile(path, strlen((char*)mJsonFileBuffer.Get()));
+    if (!err.IsNone()) {
+        return err;
+    }
+
+    return aos::ErrorEnum::eNone;
+}
+
+aos::Error OCISpec::LoadRuntimeSpec(const aos::String& path, aos::oci::RuntimeSpec& runtimeSpec)
+{
+    if (runtimeSpec.mVM == nullptr) {
+        return AOS_ERROR_WRAP(aos::ErrorEnum::eInvalidArgument);
+    }
+
+    aos::LockGuard lock(mBufferMutex);
+
+    auto redRet = ReadFileContentToBuffer(path, cJsonMaxContentSize);
+    if (!redRet.mError.IsNone()) {
+        return redRet.mError;
+    }
+
+    int ret = json_obj_parse((char*)mJsonFileBuffer.Get(), redRet.mValue, RuntimeSpecJsonDescr,
+        ARRAY_SIZE(RuntimeSpecJsonDescr), &mRuntimeSpec);
+    if (ret < 0) {
+       printk("LoadRuntimeSpec err %d \n",ret);
+    }
+
+    runtimeSpec.mVersion = mRuntimeSpec.ociVersion;
+    runtimeSpec.mVM->mHypervisor.mPath = mRuntimeSpec.vm.hypervisor.path;
+    runtimeSpec.mVM->mKernel.mPath = mRuntimeSpec.vm.kernel.path;
+
+    for (auto i = 0; i < mRuntimeSpec.vm.hypervisor.paramsLen; i++) {
+        runtimeSpec.mVM->mHypervisor.mParameters.PushBack(mRuntimeSpec.vm.hypervisor.parameters[i]);
+    }
+
+    for (auto i = 0; i < mRuntimeSpec.vm.kernel.paramsLen; i++) {
+        runtimeSpec.mVM->mKernel.mParameters.PushBack(mRuntimeSpec.vm.kernel.parameters[i]);
+    }
+
+    return aos::ErrorEnum::eNone;
+}
+
+aos::Error OCISpec::SaveRuntimeSpec(const aos::String& path, const aos::oci::RuntimeSpec& runtimeSpec)
+{
+    if (runtimeSpec.mVM == nullptr) {
+        return AOS_ERROR_WRAP(aos::ErrorEnum::eInvalidArgument);
+    }
+
+    aos::LockGuard lock(mBufferMutex);
+
+    mRuntimeSpec.ociVersion = const_cast<char*>(runtimeSpec.mVersion.CStr());
+    mRuntimeSpec.vm.hypervisor.path = const_cast<char*>(runtimeSpec.mVM->mHypervisor.mPath.CStr());
+    mRuntimeSpec.vm.kernel.path = const_cast<char*>(runtimeSpec.mVM->mKernel.mPath.CStr());
+    mRuntimeSpec.vm.hypervisor.paramsLen = runtimeSpec.mVM->mHypervisor.mParameters.Size();
+    mRuntimeSpec.vm.kernel.paramsLen = runtimeSpec.mVM->mKernel.mParameters.Size();
+
+    for (size_t i = 0; i < runtimeSpec.mVM->mHypervisor.mParameters.Size(); i++) {
+        mRuntimeSpec.vm.hypervisor.parameters[i]
+            = const_cast<char*>(runtimeSpec.mVM->mHypervisor.mParameters[i].CStr());
+    }
+
+    for (size_t i = 0; i < runtimeSpec.mVM->mKernel.mParameters.Size(); i++) {
+        mRuntimeSpec.vm.kernel.parameters[i] = const_cast<char*>(runtimeSpec.mVM->mKernel.mParameters[i].CStr());
+    }
+
+    mRuntimeSpec.vm.hwconfig.devicetree = const_cast<char*>(""); 
```

Comment:
strcpy or memset ?

Reply:
cant doit because devicetree is char* type